crypto/params.c

```diff
@@ -0,0 +1,371 @@
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+#include "internal/thread_once.h"
+
+#define SET_RETURN_SIZE(p, sz) \
+    if (p->return_size != NULL) \
+        *p->return_size = (sz)
+
+const OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->key != NULL; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+static OSSL_PARAM ossl_param_construct(const char *key, unsigned int data_type,
+                                       void *buffer, size_t buffer_size,
+                                       size_t *return_size)
+{
+    OSSL_PARAM res;
+
+    res.key = key;
+    res.data_type = data_type;
+    res.buffer = buffer;
+    res.buffer_size = buffer_size;
+    res.return_size = return_size;
+    return res;
+}
+
+#define PARAM_INT_GET(to, from) \
+    case sizeof(from): \
+        if (sizeof(to) >= sizeof(from)) { \
+            *val = (to)*(const from *)p->buffer; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT_SET(to, from) \
+    case sizeof(to): \
+        SET_RETURN_SIZE(p, sizeof(to)); \
+        if (sizeof(to) >= sizeof(from)) { \
+            *(to *)p->buffer = (to)val; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT(name, type, pre) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, type *val) \
+    { \
+        if (val == NULL \
+            || p == NULL \
+            || (p->data_type != OSSL_PARAM_INTEGER \
+                && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)) \
+            return 0; \
+        switch (p->buffer_size) { \
+        PARAM_INT_GET(type, pre ## 32_t); \
+        PARAM_INT_GET(type, pre ## 64_t); \
+        } \
+        return 0; \
+    } \
+    int OSSL_PARAM_set_##name(const OSSL_PARAM *p, type val) \
+    { \
+        if (p == NULL) \
+            return 0; \
+        SET_RETURN_SIZE(p, 0); \
+        if (p->data_type == OSSL_PARAM_INTEGER \
+            || p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) \
+            switch (p->buffer_size) { \
+            PARAM_INT_SET(pre ## 32_t, type); \
+            PARAM_INT_SET(pre ## 64_t, type); \
+            } \
+        return 0; \
+    } \
+    OSSL_PARAM OSSL_PARAM_construct_##name(const char *key, type *buf, \
+                                           size_t *rsize) \
+    { \
+        return ossl_param_construct(key, #pre[0] == 'u' \
+                                         ? OSSL_PARAM_UNSIGNED_INTEGER \
+                                         : OSSL_PARAM_INTEGER, \
+                                    buf, sizeof(type), rsize); \
+    }
+
+PARAM_INT(int, int, int)
+PARAM_INT(uint, unsigned int, uint)
+PARAM_INT(long, long int, int)
+PARAM_INT(ulong, unsigned long int, uint)
+PARAM_INT(int32, int32_t, int)
+PARAM_INT(uint32, uint32_t, uint)
+PARAM_INT(int64, int64_t, int)
+PARAM_INT(uint64, uint64_t, uint)
+PARAM_INT(size_t, size_t, uint)
```

Comment:
Frankly, I'd rather all of the above was not done with macro trickery. It makes it very difficult to read and it will not be possible to debug such code. Also searching the code trying to find the definition of `OSSL_PARAM_construct_uint32` (or whatever) becomes much harder.

Reply:
The macros have been expanded inline and simplified in some cases.